tmt/steps/provision/__init__.py

```diff
@@ -449,32 +449,35 @@ def _ansible_playbook_path(self, playbook):
         self.debug(f"Playbook full path: '{playbook}'", level=2)
         return playbook
 
-    def _export_environment(self, execute_environment=None):
-        """ Prepare shell export of environment variables """
+    def _prepare_environment(self, execute_environment=None):
+        """ Prepare dict of environment variables """
         # Prepare environment variables so they can be correctly passed
         # to ssh's shell. Create a copy to prevent modifying source.
         environment = dict()
         environment.update(execute_environment or dict())
         # Plan environment and variables provided on the command line
         # override environment provided to execute().
         environment.update(self.parent.plan.environment)
-        # Prepend with export and run as a separate command.
-        if not environment:
-            return ''
-        return 'export {}; '.format(
-            ' '.join(tmt.utils.shell_variables(environment)))
+        return {} if not environment else environment
+
+    @staticmethod
+    def _export_environment(environment):
+        """ Prepare shell export of environment variables """
+        return f'export {" ".join(tmt.utils.shell_variables(environment))}; ' \
+            if environment \
+            else ''
 
     def ansible(self, playbook, extra_args=None):
         """ Prepare guest using ansible playbook """
         playbook = self._ansible_playbook_path(playbook)
         stdout, stderr = self.run(
-            f'{self._export_environment()}'
-            f'stty cols {tmt.utils.OUTPUT_WIDTH}; ansible-playbook '
-            f'--ssh-common-args="{self._ssh_options(join=True)}" '
-            f'{self._ansible_verbosity()} '
-            f'{self._ansible_extra_args(extra_args)} -i {self._ssh_guest()},'
-            f' {playbook}',
-            cwd=self.parent.plan.worktree)
+            ['ansible-playbook'] +
+            shlex.split(self._ansible_verbosity() + ' ' +
+                        self._ansible_extra_args(extra_args)) +
+            [*(f'--ssh-common-args="{i}"' for i in self._ssh_options()),
```

Comment:
Why are you packing and unpacking the common args here? Why not use just list comprehension?

```python
[f'--ssh-common-args="{i}"' for i in self._ssh_options()]
```

The current way is a bit confusing to me.

Reply:
Sorry, it was an intermediate version. Could be done this way:
```
            ['ansible-playbook'] +
            shlex.split(self._ansible_verbosity() + ' ' +
                        self._ansible_extra_args(extra_args)) +
            [f'--ssh-common-args="{i}"' for i in self._ssh_options()] +
            ['-i', f'{self._ssh_guest()},', playbook],
```

Reply:
Looks much better to me. Last thing: The grouping of verbosity and extra args is to handle empty verbosity? Let's make that rather explicit. Perhaps something like this?

```yaml
[verbosity] if verbosity else []
```

Reply:
> is to handle empty verbosity?
Correct